test/e2e/unmanaged_dns_test.go

```diff
@@ -0,0 +1,380 @@
+		t.Fatalf("failed to delete svc %s: %v", lbService.Name, err)
+	}
+
+	t.Log("Waiting for stable conditions on ingresscontroller after Managed")
+
+	// Wait for the load balancer and DNS to reach stable conditions.
+	if err := waitForIngressControllerCondition(t, kclient, 10*time.Minute, name, availableConditionsForIngressControllerWithLoadBalancer...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	if !ingresscontroller.IsServiceInternal(lbService) {
+		t.Fatalf("load balancer %s is internal but should be external", lbService.Name)
+	}
+
+	if err := kclient.Get(context.TODO(), wildcardRecordName, wildcardRecord); err != nil {
+		t.Fatalf("failed to get wildcard dnsrecord %s: %v", wildcardRecordName, err)
+	}
+
+	if wildcardRecord.Spec.DNSManagementPolicy != iov1.ManagedDNS {
+		t.Fatalf("DNSRecord %s expected in ManagedDNS state but got %s", wildcardRecordName.Name, wildcardRecord.Spec.DNSManagementPolicy)
+	}
+
+	if len(wildcardRecord.Status.Zones) != 2 {
+		t.Fatalf("DNSRecord %s expected allocated dnsZones but found none", wildcardRecordName.Name)
+	}
+
+	verifyUnmanagedIngressController(t, name, "apps."+ic.Spec.Domain, wildcardRecord.Spec.Targets[0])
+}
+
+func verifyUnmanagedIngressController(t *testing.T, name types.NamespacedName, hostname, address string) {
+	echoPod := buildEchoPod(name.Name, name.Namespace)
+	if err := kclient.Create(context.TODO(), echoPod); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", echoPod.Namespace, echoPod.Name, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), echoPod); err != nil {
+			t.Fatalf("failed to delete pod %s/%s: %v", echoPod.Namespace, echoPod.Name, err)
+		}
+	}()
+
+	echoService := buildEchoService(echoPod.Name, echoPod.Namespace, echoPod.ObjectMeta.Labels)
+	if err := kclient.Create(context.TODO(), echoService); err != nil {
+		t.Fatalf("failed to create service %s/%s: %v", echoService.Namespace, echoService.Name, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), echoService); err != nil {
+			t.Fatalf("failed to delete service %s/%s: %v", echoService.Namespace, echoService.Name, err)
+		}
+	}()
+
+	echoRoute := buildRouteWithHost(echoPod.Name, echoPod.Namespace, echoService.Name, hostname)
+	if err := kclient.Create(context.TODO(), echoRoute); err != nil {
+		t.Fatalf("failed to create route %s/%s: %v", echoRoute.Namespace, echoRoute.Name, err)
+	}
+	defer func() {
+		if err := kclient.Delete(context.TODO(), echoRoute); err != nil {
+			t.Fatalf("failed to delete route %s/%s: %v", echoRoute.Namespace, echoRoute.Name, err)
+		}
+	}()
+
+	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("http://%s", address), nil)
+	if err != nil {
+		t.Fatalf("failed to build client request: %v", err)
+	}
+	req.Host = hostname
+
+	httpClient := http.Client{Timeout: 5 * time.Second}
+	err = waitForHTTPClientCondition(t, &httpClient, req, 10*time.Second, 5*time.Minute, func(r *http.Response) bool {
```

Comment:
Why does `verifyUnmanagedIngressController` build a pod, service, route and try to curl it? I guess it's not super clear to me how this tests DNS.

Reply:
So the route that is created is configured to use the created ingresscontroller's domain. The curl tries to ensure proper connectivity to the test workload (echo pod) and this also ensures functionality of the router that was created.

Reply:
That makes sense and is valid. Did you consider using just a `dnslookup` instead of HTTP client request? An entire http request makes sense, but seems to be quite broad. Just curious, not necessarily recommending a change.

My other comment is, now that I understand these functions, would it make more sense to put them in the util_test.go? Because these are really quite generic IngressController tests that could be reused.

Reply:
> That makes sense and is valid. Did you consider using just a dnslookup instead of HTTP client request? An entire http request makes sense, but seems to be quite broad. Just curious, not necessarily recommending a change.

I initially did think of using `dnslookup`, but found a couple of issues with it.
- In the scenario where we have a `Unmanaged` external ingresscontroller, we would need to create new DNS records on the cloud providers (and their cleanup). This kind of adds a lot of test code to maintain.
- Since the test creates a lot of new objects (router,certs,secrets,ingresscontrollers) just testing using `dnslookup` doesn't exactly cover all aspects. So adding a HTTP call, ensures a broad test covering all the new objects.

> My other comment is, now that I understand these functions, would it make more sense to put them in the util_test.go? Because these are really quite generic IngressController tests that could be reused.

Agreed, moved to `util_test.go` along with `godocs`.